src/agent/src/watcher.rs

```diff
@@ -0,0 +1,779 @@
+            if !entry.watch {
+                continue;
+            }
+
+            match entry.scan(logger).await {
+                Err(error) => {
+
+                    // If an error was observed, we will stop treating this WatchedStorage as being watchable, and
+                    // instead clean up the target-mount files on the tmpfs and bind mount the source_mount_point to target_mount_point.
+                    error!(logger, "error observed when watching: {:?}", error);
+                    entry.watch = false;
+
+                    // Remove destination contents, but not the directory itself, since this is
+                    // assumed to be bind-mounted into a container. If source/mount is a file, no need to cleanup
+                    if entry.target_mount_point.as_path().is_dir() {
+                        for dir_entry in std::fs::read_dir(entry.target_mount_point.as_path())? {
+                            let dir_entry = dir_entry?;
+                            let path = dir_entry.path();
+                            if dir_entry.file_type()?.is_dir() {
+                                tokio::fs::remove_dir_all(path).await?;
+                            } else {
+                                tokio::fs::remove_file(path).await?;
+                            }
+                        }
+                    }
+
+                    //  - Create bind mount from source to destination
+                    BareMount::new(
+                        entry.source_mount_point.to_str().unwrap(),
+                        entry.target_mount_point.to_str().unwrap(),
+                        "bind",
+                        MsFlags::MS_BIND,
+                        "bind",
+                        logger,
+                    )
+                    .mount()?;
+                }
+                Ok(count) => {
+                    debug!(logger, "watched {} files", count);
+                }
+            };
+        }
+
+        Ok(())
+    }
+}
+
+/// Handles watchable mounts. The watcher will manage one or more mounts for one or more containers. For each
+/// mount that is added, the watcher will maintain a list of files to monitor, and periodically checks for new,
+/// removed or changed (modified date) files. When a change is identified, the watcher will either copy the new
+/// or updated file to a target mount point, or remove the removed file from the target mount point.  All WatchableStorage
+/// target mount points are expected to reside within a single tmpfs, whose root is created by the BindWatcher.
+///
+/// This is a temporary workaround to handle config map updates until we get inotify on 9p/virtio-fs.
+/// More context on this:
+/// - https://github.com/kata-containers/runtime/issues/1505
+/// - https://github.com/kata-containers/kata-containers/issues/1879
+#[derive(Debug, Default)]
+pub struct BindWatcher {
+    /// Container ID -> Vec of watched entries
+    shared: Arc<Mutex<HashMap<String, Entries>>>,
```

Comment:
I'm really picking on naming, sorry about that. Using `shared` makes the code below harder to read, while renaming it to e.g `containers_storage` would make e.g. the following code:

```Rust
self.containers_storage
            .lock()
            .await
            .entry(id)
            .or_insert_with(ContainerStorage::default)
            .add(mounts, logger)
            .await
            .with_context(|| "Failed to add container")?;
```

easier to understand/read.